addons/product_images/wizard/product_fetch_image_wizard.py

```diff
@@ -0,0 +1,211 @@
+        nb_product_processable = len(product_processable)
+        nb_product_unprocessable = nb_product_selected - nb_product_processable
+
+        default['product_processable'] = product_processable
+        # Not computed because the value of product_processable doesn't exist yet.
+        default['nb_product_selected'] = nb_product_selected
+        default['nb_product_processable'] = nb_product_processable
+        default['nb_product_unprocessable'] = nb_product_unprocessable
+
+        return default
+
+    product_processable = fields.Many2many(
+        comodel_name='product.product',
+        help="List of the selected product without a picture and with a barcode",
+    )
+    nb_product_selected = fields.Integer(string="Number of product selected", readonly=True)
+    nb_product_processable = fields.Integer(string="Number of product processable", readonly=True)
+    nb_product_unprocessable = fields.Integer(string="Number of product unprocessable",
+                                              readonly=True)
+
+    def action_fetch_image(self):
+        """ Trigger a cron job to fetch image on Google
+
+        The 10 first images are immediately fetched to improve the user experience. This way, they
+        can immediately view the images and be assured that the task is running well. Also, if any
+        error occurs, they can be thrown to the user. Then, a cron job is triggered to be run as
+        soon as possible, unless the daily request limit has been reached. In that case, the cron
+        job will be scheduled to run a day later.
+
+        :raises UserError: If the project is misconfigured on Google's side
+        :raises UserError: If the API Key or Search Engine ID is incorrect
+        :return: A notification to inform the user that the task is triggered
+        :rtype: `ir.actions.client`
+        """
+        self.product_processable.image_fetch_pending = True
+        try:
+            self._fetch_image_from_google(self.product_processable[:10])
+        except HTTPError as http_error:
+            if http_error.response.status_code == requests.codes.too_many_requests:
+                self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                    fields.Datetime.now() + timedelta(days=1.0)
+                )
+                return {
+                    'type': 'ir.actions.client',
+                    'tag': 'display_notification',
+                    'params': {
+                        'type': 'warning',
+                        'title': _("Product images"),
+                        'message': _("Custom search quota exceeded. We will start again tomorrow"),
+                        'next': {'type': 'ir.actions.act_window_close'},
+                    }
+                }
+            elif http_error.response.status_code == requests.codes.forbidden:
+                project_url = re.search(r"(?P<url>https?://[^\s]+)",
```

Comment:
I have never seen this notation for capturing groups. Is the `P` making any change?

Reply:
Actually, do you even need a group here, since you're capturing the entire expression?

Reply:
It seems that it is part of the syntax of python's re module. If I remove the 'P', the syntax should still be correct but I receive a traceback in Odoo. I based this regex on [this document](https://www.regular-expressions.info/named.html) since I never done one in Python before (and never used the 'P' either). The group allows me to fetch directly the URL at the next line: https://github.com/odoo/odoo/blob/8d6bf51a699dbfbdc373c8a044fdf79bf412edbf/addons/product_images/wizard/product_fetch_image_wizard.py#L99

Reply:
Ok, fine by me! I still believe that the URL can be obtained from the returned `match` object but it's not a big deal here.